client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/XaModeTransactionTable.java

```diff
@@ -208,121 +250,44 @@ private void assertSameXid(Xid otherXid, int xaErrorCode) throws XAException {
-            if (alreadyForgot) {
-               //no tx
-               throw new XAException(XAException.XAER_NOTA);
-            } else if (unknown) {
-               //unknown. we throw what we received
-               throw new XAException(unknownErrorCode);
-            }
-         } else if (hasCommit && hasRollback) {
-            //we have something committed and something rolled back
-            throw new XAException(XAException.XA_HEURMIX);
-         } else if (hasCommit) {
-            //we only got commits
-            throw new XAException(XAException.XA_HEURCOM);
-         }
-         //we only got rollbacks
+      private TransactionContext<?, ?> getAnyTransactionContext() {
+         return registeredCaches.values().iterator().next();
       }
 
       private void internalCommit() throws XAException {
-         boolean hasCommit = false;
-         boolean hasRollback = false;
-         boolean alreadyForgot = false;
-         boolean unknown = false;
-         int unknownErrorCode = 0;
-
-         for (TransactionContext<?, ?> ctx : preparedCaches) {
-            int xa_code = ctx.complete(currentXid, true);
-            switch (xa_code) {
-               case XAResource.XA_OK:       //no issues
-               case XAResource.XA_RDONLY:   //no issues
-               case XAException.XA_HEURCOM: //heuristically committed
-                  hasCommit = true;
-                  break;
-               case XAException.XAER_NOTA:
-                  //tx already forgotten by the server.
-                  alreadyForgot = true;
-                  break;
-               case XAException.XA_HEURMIX: //completed. not sure how
-               case XAException.XA_HEURHAZ: //heuristically committed and rolled back
-                  hasCommit = true;
-                  hasRollback = true;
-                  break;
-               case XAException.XA_HEURRB: //heuristically rolled back
-                  hasRollback = true;
-                  break;
-               default:
-                  //other error codes
-                  unknown = true;
-                  unknownErrorCode = xa_code;
-                  break;
-            }
-         }
-
-         if (!hasCommit && !hasRollback) {
-            //nothing committed neither rolled back
-            if (alreadyForgot) {
-               //no tx
-               throw new XAException(XAException.XAER_NOTA);
-            } else if (unknown) {
-               //unknown. we throw what we received
-               throw new XAException(unknownErrorCode);
-            }
-         } else if (hasCommit && hasRollback) {
-            //we have something committed and something rolled back
-            throw new XAException(XAException.XA_HEURMIX);
-         } else if (hasRollback) {
-            //rollbacks only
-            throw new XAException(XAException.XA_HEURRB);
+         int xa_code = getAnyTransactionContext().complete(currentXid, true);
```

Comment:
It might be cleaner if the operations were executed at the adapter level and not at the `TransactionContext` level. Collecting the modifications is done per cache, but the modifications list can be a parameter of `prepareContext()`.

Reply:
I had the operation in adapter before and I end up duplicate the code between SyncAdapter and XaAdapter. So I moved them to the context

Reply:
ok, I'll change for CR1